### ISSUE TITLE
Refine the definition of Distinctive Identifier

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#117]
This issue tracks finalization of the [definition of Distinctive Identifier](https://w3c.github.io/encrypted-media/#distinctive-identifier), including addressing the remaining issues from [Bug 27268](https://www.w3.org/Bugs/Public/show_bug.cgi?id=27268).

**April 15, 2016 Update:**
* https://github.com/w3c/encrypted-media/issues/117#issuecomment-210692963 summarizes the problems with the current definition of Distinctive Identifier and explores what it should address.
* https://github.com/w3c/encrypted-media/issues/117#issuecomment-210698890 covers the remaining issues from [Bug 27268](https://www.w3.org/Bugs/Public/show_bug.cgi?id=27268).

**Comment by steelejoe** [comment]
@ddorwin do you have a first cut of the list of issues you can put here for discussion? This is likely to be a long/deep conversation and I would like to help.

**Comment by ddorwin** [comment]
Hi @steelejoe. I have started work on this, but it's not quite there yet.

Since this topic is already complex, please file a separate issue for discussion of that sentence from 8.1.2.

**Comment by ddorwin** [comment]
In addition to the remaining issues from the original bug (see the next comment), the [current definition of Distinctive Identifier](https://cdn.rawgit.com/w3c/encrypted-media/50ebd0cdd9b3f60529707a6f2560d748e74c443e/index.html#distinctive-identifier) is overly broad. For example, it treats a truly random value, most persistent state, and even a cookie the same as a value based on a permanent hardware identifier (as [discussed in the original bug](https://www.w3.org/Bugs/Public/show_bug.cgi?id=27268#c6)). This is inconsistent with the intent, waters down the meaning, and reduces incentives and advantages of avoiding the latter. We need to correct this while still providing the user a level of privacy that is as consistent as practical with the rest of the web platform. The following is a summary of my current thinking. I am working on how to address this in the spec text.


<br/>There are somewhat conflicting goals: Require identifiers used by the Key System implementation to be like cookies and other web storage (unique per origin, clearable, non-associable, etc.) while still allowing "reasonable" functionality, such as attestation, as long as the user consents.

As with all identifiers, I believe the concern is misuse, especially tracking. While any reasonably stable identifier will allow a site to track the user until he or she clears it (cookies already support this), what is more concerning is:
* The ability to track across sites (origins).
  * For example, third-party cookies (e.g. used by ad networks) and omnipresent widget iframes, which can bypass third-party cookie checks.
* The ability to track over time, including inability (or difficulty) to easily clear identifiers.

Thus, the real concern here is that someone - anyone - could somehow a) tie a Distinctive Identifier back to a specific device or b) correlate two Distinctive Identifiers from different origins or across clearing of Distinctive Identifier(s), especially after the user has made explicit efforts to prevent this, such as by clearing browsing data.

Since the EME spec already requires that identifiers and other data exposed to applications be (mostly) like cookies (unique per origin, clearable in some way, non-associable), I believe we should focus on the generation of those identifiers, including the individualization process. If the individualization entity receives a distinctive value or could otherwise potentially make a connection, the resulting value is a Distinctive Identifier.


<br/>An analogy is temporary credit card numbers. (Disclaimer: This is an approximation and may not be entirely accurate, but any inaccuracies are likely irrelevant.)
* The merchant (application/content provider) knows the customer has a Visa card (Foo Key System implementation) and perhaps which bank issued it (device class or model).
* The merchant (application/content provider) does _not_ know the real credit card number (indelible device value) and cannot correlate it with numbers used for future transactions (browsing sessions if the user clears Distinctive Identifier and other data) with the same merchant or with transactions at other merchants (other applications/content providers).
* But the credit card company (Key System provider) can correlate all of the numbers, which is necessary for its operations, such as credit card billing (client verification).
* The customer (user) controls the generation and use of both values:
 * Real credit card numbers (indelible values) are generated when the user applies for credit.
 * Temporary credit card numbers (values based on indelible values) are generated and provided upon request and authorization of the user.

Applying this to EME:
* We want to:
 * Protect and restrict access to any indelible values (real credit card numbers)
 * Ensure Distinctive Identifiers (temporary credit card numbers) exposed to content providers are not associable ("random" temporary credit card numbers)
 * Ensure users are in control of and consent to when and to whom indelible values (real credit card numbers) and values based on them (temporary credit card numbers) are exposed outside the client device.
* It is not a significant concern if the application continues to use a Distinctive Identifiers (temporary credit card number) or other persistent information (e.g. shipping address), including cookies, persistent sessions, and other persistent state, as long as the user has not attempted to clear this information (or exited a private browsing session).
* Once the user chooses to clear this information, it MUST NOT be possible to correlate identifiers provided to different applications/content providers (merchants).
* If a user uses private browsing, another profile, or another OS user account, the application (content provider or merchant) MUST NOT be able to correlate it with other browsing sessions, profiles, or accounts.

**Comment by ddorwin** [comment]
#168 and #169 address concerns about other types of data that might be used for tracking. These should address the remaining issues from the original bug, leaving this one to focus on the definition of Distinctive Identifier.

**Comment by ddorwin** [comment]
The definition should reference https://w3c.github.io/encrypted-media/#allow-identifiers-cleared and potentially other sections. The text referenced in https://github.com/w3c/encrypted-media/issues/168#issuecomment-210700176 is particularly relevant.

**Comment by ddorwin** [comment]
Reviewed in the editors meeting. We agreed with following.

Our intent on this issue is to clarify the definition of Distinctive Identifiers to exclude simple IDs like cookies from Distinctive Identifier requirements in this spec.

Identifiers exposed to the application must already always be non-associable after clearing and across origins. If it is non-associable by any entity, including an individualization server, then it is just like a cookie and Distinctive Identifier does not apply. However, if it is possible to associate the identifiers outside the device, such as by the individualization server, then the Distinctive Identifier conditions do apply. Possible causes of the latter include use of hardware or other indelible IDs in the individualization process.

**Comment by ddorwin** [comment]
I created PR #218.

**Comment by ddorwin** [comment]
Reviewing the text with the new definition, I see the following issues:

1. Other types of identifiers (i.e. those whose generation does not involve indelible value(s)) are no longer covered by various requirements, including those in the [identifier requirements](https://w3c.github.io/encrypted-media/#identifier-requirements) section.
1. Phrases like "use Distinctive Identifier(s)" do not necessarily cover a case where an indelible value is sent to an individualization server but a resulting identifier is _not_ generated.

For (1), I think we need some definition (i.e. persistent identifier) that looks like the previous Distinctive Identifier definition. We can then use it in the requirements section.

For (2), we really want to cover any time an indelible value is exposed outside the client. While it might seem that the obvious answer is to add "it is indelible" to the possible properties that makes a Distinctive Identifiers, such a value would not be clearable or per-origin as required elsewhere.

One possible solution is to define a term (i.e. Permanent Identifier) for the indelible values that are among the conditions that make a Distinctive Identifiers then explicitly reference that in the various "use Distinctive Identifier(s)" steps. That might involve defining a new term - "Use Distinctive Identifier [or Permanent Identifier]", which could go explicitly cover these issues.

**Comment by ddorwin** [comment]
Other thoughts:
* (1) should be a superset of Distinctive Identifiers.
* Some of the definition for (1) will be duplicated in the Distinctive Identifier definition. It might be nice to say a  Distinctive Identifier is a (1) with these additional properties, but there may be a circular reference due to the previous bullet, which is more important.
* Most references to Distinctive Identifier in sections 10 and 11 should probably be changed to (1).

**Comment by ddorwin** [comment]
"Distinctive Identifier" would probably accurately describe (1) above, but we need to ensure that the "Distinctive Identifier" definition matches the intended behavior of the `distinctiveIdentifier` member in the API since - unlike the rest of the uses in the spec - this use of the term cannot be changed.

I had always assumed that "Distinctive Identifier" and the associated requirements referred to the more concerning type of identifier described by PR #218 and that `distinctiveIdentifier` would thus refer to such identifiers. However, looking at the [Private Browsing Modes](https://w3c.github.io/encrypted-media/#private-browsing) text has me rethinking this.

Allowing private browsing and other stateless modes/implementations to deny such functionality was one of the reasons `distinctiveIdentifier` and `persistentState` were originally added. It would seem that such modes might also want to reject (1). With #218, it is not possible for an application to determine whether (1) is supported.

On the other hand, maybe it is more important for applications to be able to query support for the concerning type and thus we should continue to use the definition from #218. This is probably the more practical use. Even if private browsing modes do not allow (1), this may have minimal relevance to an application in practice.

It's possible we have been conflating the privacy properties of identifiers related to indelible values with the use of any type of identifier. Perhaps one way to look at this is to ask how should an implementation that uses a semi-persistent random value behave. Note that cookie-like identifiers would continue to be controlled by `persistentState`. Another option would be to treat truly random identifiers - and perhaps other instances of (1) - as `persistentState`, since (1) must be clearable "along with cookies and other site data."

**Comment by ddorwin** [comment]
Related to the above, in https://github.com/w3c/encrypted-media/issues/117#issuecomment-222186131, we said, "Our intent on this issue is to... exclude simple IDs like cookies from Distinctive Identifier requirements in this spec." The specific solution we identified was to "clarify the definition of Distinctive Identifiers to" exclude them (this was the "..."), but that is not the only option.

**Comment by ddorwin** [comment]
PR #228 adds additional clarification to the current definition (#218) and two ISSUE boxes for the issues identified in https://github.com/w3c/encrypted-media/issues/117#issuecomment-224129928.

**Comment by ddorwin** [comment]
I propose the following changes to address the comments yesterday and the two ISSUE boxes added in #228:

1. Leave the meaning of `distinctiveIdentifier` and Distinctive Identifiers as is.
1. Define Permanent Identifier
      - This covers indelible values and similar whose use makes an identifier a Distinctive Identifier.
      - Move the related text from the Distinctive Identifier definition and refer to the new definition from there.
      - Permanent Identifiers MUST NOT ever be exposed to the application, even in encrypted for.
1. Define Use Distinctive Identifier or Permanent Identifier and reference it from the algorithms that test the implementation for a configuration and prevent use.
1. Break up the [identifier requirements](https://w3c.github.io/encrypted-media/#identifier-requirements) section.
   - The first two subsections (limit use and decrypt) apply to Distinctive Identifiers.
   - The last two subsections (per-origin, clearable) apply to all identifiers (and any other potentially-identifying information exposed to the application).
      - This includes random values, cookies, and session data, and other persistent data.
1. Update sections 10 (Security) and 11 (Privacy) as appropriate.
   - This may include adding "or other identifiers" to some uses of "Distinctive Identifiers."

**Comment by steelejoe** [comment]
The pull request looks good so far. I believe you have addressed all of my concerns with Distinctive Identifier.

**Comment by ddorwin** [comment]
PR #228 was merged as https://github.com/w3c/encrypted-media/commit/7dcdc3317965a95b6b138e175a132b0b7d46a7a4.

PR #230 clarifies the meaning of "associable" for Distinctive Identifiers.

**Comment by mwatson2** [comment]
Your [proposal](https://github.com/w3c/encrypted-media/issues/117#issuecomment-224432229) looks good to me. Regarding private browsing modes, I think what data is kept across browsing sessions within a single instance of the private browsing mode and what data is deleted when you leave private browsing mode are implementation issues, since such modes are not presently standardized.

Use of identifiers that are not Distinctive Identifiers (i.e. that have the same properties as Cookies) does not need to be exposed to or controlled by applications in the same way as Distinctive Identifiers.

**Comment by ddorwin** [comment]
I have started work related to defining "Permanent Identifier." This will be ready tomorrow morning PDT.

**Comment by ddorwin** [comment]
I think we may need to use "**Distinctive** Permanent Identifier" since a permanent value "shared across a large population of users or client devices" should not trigger the related behaviors. Other naming suggestions welcome.

**Comment by ddorwin** [comment]
PR #238 adds Distinctive Permanent Identifier and Use of Distinctive Identifiers and Distinctive Permanent Identifiers.

**Comment by ddorwin** [comment]
PR #241 to cover (4) in my [proposal](https://github.com/w3c/encrypted-media/issues/117#issuecomment-224432229). There was no consistent way to categorize the requirements, so I left them in the same structure but summarized how they apply.

**Comment by ddorwin** [comment]
Regarding the items in my [proposal](https://github.com/w3c/encrypted-media/issues/117#issuecomment-224432229) above:

1. NOP
1. PR #238 defined Permanent Identifier and Distinctive Permanent Identifier.
1. PR #238 defined use of Distinctive Identifiers and and Distinctive Permanent Identifiers.
1. PR #241 updated the identifier requirements section.
1. Addressed by PR #238 and maybe others. I filed #237 to revisit the User Tracking section to address Permanent Identifiers.

I believe this is now fixed.
